Introduction to Anypoint Connectors topic (#518)
mulesoft/docs-connectors: modules/ROOT/pages/introduction/introduction-to-anypoint-connectors.adoc

= Introduction to Anypoint Connectors
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

Anypoint Connectors are reusable extensions of Mule that enable you to integrate a Mule app with third-party APIs and standard integration protocols. All connectivity in Mule 4 is provided through connectors.

Using connectors in an app provides provides the following advantages:

* Simplifies app development and saves time, since you don't have to program directly to APIs
* Can be updated without requiring updates to other parts of the app

> * *Reviewers, are there other advantages I can add here?*

Reviewer: Maybe also this one:
<img width="444" alt="image" src="https://user-images.githubusercontent.com/23425672/70911078-68065600-2011-11ea-986b-132760371fce.png">

To see a list of operations for a connector, see the Reference Guide for that connector.

== Access Connectors and Related Assets

Anypoint Exchange provides access to all publicly-available connector assets including connectors, templates, and examples.

=== Connectors in Exchange

You can use Exchange as a starting point for discovering all or a subset of MuleSoft-provided connectors:

For information about locating connectors in Exchange, see <<introduction/discover-add-connectors.adoc#Discover and Add Connectors>>.

=== Templates in Exchange

Templates provide starting points for creating Studio projects. Many templates include one or more connector configurations.  For example, the S/4HANA Cloud to Salesforce Partner Migration template helps you migrate partners from an SAP S/4HANA Cloud to a Salesforce instance.

> Once you configure a template in Studio, you can customize it to meet your needs. *Reviewers, is this true in Flow Designer as well?*

Reviewer: Yes it is, please see https://docs.mulesoft.com/design-center/import-template

= Introduction to Anypoint Connectors
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

> Anypoint Connectors are reusable extensions of Mule that enable you to integrate a Mule app with third-party APIs, databases, and standard integration protocols. Connectors abstract the technical details involved with connecting to a target system. All connectivity in Mule 4 is provided through connectors.

Reviewer 2: Does this sound better? I want to use the full name of Mule right up front:
```suggestion
Anypoint Connectors are reusable extensions to Mule runtime engine (Mule) that enable you to integrate a Mule app with third-party APIs, databases, and standard integration protocols. Connectors abstract the technical details involved in connecting to a target system. All connectivity in Mule 4 is provided through connectors.
```

Author: Is this a recommendation to use the term Mule runtime engine on first mention? I am OK with it - just haven't heard it before.

Reviewer 2: Yes, please.

> Using connectors in an app provides provides the following advantages:

Reviewer 2: remove duplicate word:
```suggestion
Using connectors in an app provides the following advantages:
```
Should "an app" be "a Mule app"?

Reviewer 2: Remove periods from list items below, which are not complete sentences.

> * Reduces code complexity, because you can connect a Mule app to a target system without knowing all of the details required to program to the target system.

Reviewer 2: ```suggestion
* Reduces code complexity, because you can connect a Mule app to a target system without knowing all of the details required to program to the target system
```

> * Simplifies authenticating against the target system.

Reviewer 2: ```suggestion
* Simplifies authenticating against the target system
```

> * Proactively infers metadata for the target system, which makes it easier to identify and transform data with DataWeave.

Reviewer 2: ```suggestion
* Proactively infers metadata for the target system, which makes it easier to identify and transform data with DataWeave
```

* Makes code maintenance easier because:
** Not all changes in the target system require changes to the app.
** The connector configuration can be updated without requiring updates to other parts of the app.

> == Enable Connectivity with Connectors

Reviewer 2: Your opinion, please:
Do you think that it is worthwhile to create a style recommendation to reserve imperative (enable, connect) to topics that are actual tasks, with steps, and to use the gerund form for those that do not (enabling, connecting)? I asked about this when I first began in this dept., but was overwhelmed by other issues. Now I want to return to it with a little experience behind me. Just your opinion.

Author: I have always liked gerunds and do not understand why Salesforce recommends not using them. At my last job, we did what you are suggesting - used gerunds for conceptual material and imperative for tasks.

Reviewer 2: Thanks. I'll keep banging away at trying to get this changed.

Reviewer 2: I think it's a matter of thinking that -ing and -ion are not direct enough?

Use connectors to connect a Mule app to specific software applications, databases, and protocols. To see a list of the MuleSoft-built connectors in Mule 4, see xref:index.adoc[About Connectors and Modules (for Mule 4)].

=== Connect to Software Applications

You can use connectors to connect a Mule app to specific software applications and to perform actions on the connected application.

> For example, you can use Anypoint Connector for SAP (SAP Connector) to automate sales order processing between SAP ERP and your CRM.

Reviewer 2: Will our audience understand all of the abbreviations used in this sentence?

> Likewise, you can use Anypoint Connector for Salesforce (Salesforce Connector) to integrate Salesforce with other business applications such as ERP, analytics, and data warehouse systems.

Reviewer 2: ```suggestion
Likewise, you can use Anypoint Connector for Salesforce (Salesforce Connector) to integrate Salesforce with other business applications, such as ERP, analytics, and data warehouse systems.
```
OR
```suggestion
Likewise, you can use Anypoint Connector for Salesforce (Salesforce Connector) to integrate Salesforce with business applications such as ERP, analytics, and data warehouse systems.
```

=== Connect to Databases

You can use connectors to connect a Mule app to one or more databases and to perform actions on the connected database.

For example, you can use Anypoint Connector for Databases (Database Connector) to connect a Mule app to any relational database engine. Then you can perform SQL queries on that database.

Likewise, you can use Anypoint Connector for Hadoop Distributed File System (HDFS Connector) to connect a Mule app to a Hadoop Distributed File System (HDFS). Then you can integrate databases such as MongoDB with Hadoop file systems to read, write, receive files, and send files on the HDFS server.

=== Connect to Protocols

You can use connectors to send and receive data over protocols and, for some protocol connectors, to perform protocol operations.

> For example, you can use Anypoint Connector for LDAP (LDAP Connector) to connect to an LDAP server and access Active Directory. Then you can add user accounts, delete user accounts, or retrieve user attributes, such as the user's e-mail or phone number.

Reviewer 2: e-mail > email
GLOBAL

Likewise, you can use Anypoint Connector for WebSockets (WebSockets Connector) to establish a WebSocket for bidirectional and full-duplex communication between a server and client, and to implement server push notifications.

== How Connectors Work

Connectors can perform one or more functions in an app, depending on where you place them and the capabilities of the specific connector. A connector can act as:

> * An inbound endpoint that starts the app flow. Connectors that have <<input-sources,input sources>> can perform this function.

Reviewer 2: GLOBALLY limit list items and steps to one sentence (or fragment) max:
```suggestion
* An inbound endpoint that starts the app flow
Connectors that have <<input-sources,input sources>> can perform this function.
```

* A message processor that performs <<operations,operations>> in the middle of a flow
* An outbound endpoint that receives the final payload data from the flow

[[input-sources]]
=== Input Sources

> Some connectors have input sources, or "triggers". These components enable a connector to start a Mule flow by receiving information from the connector’s resource. For example, when a Salesforce user updates a sales opportunity, a flow that starts with a Salesforce Connector component can receive the information and direct it to a database connector for processing.

Reviewer 2: I'd also like to ask for your opinion about introducing and then using synonyms.
If we intro "input sources" and then call them "triggers," should we then use "triggers" consistently in the following text, or should we use "input sources" only, because it is the first name mentioned? Or does it not matter?

Author: I would use the first term mentioned only, unless we are referencing the term "trigger" in the UI. In this case, the Flow Designer GUI uses the term "trigger" but the term in the industry is "input sources." However, this is a good question, especially for the Using Flow Designer to Configure a Connector topic.

Reviewer 2: Should we take this up in the team channel?

To see if a connector can act as an input source, see the Reference Guide for the connector.

App developers can also use an HTTP Listener or Scheduler as an input source for a flow:

> * HTTP Listener is a connector that listens to HTTP requests. You can configure an HTTP Listener to start a flow when it receives specified requests.

Reviewer 2: ```suggestion
* HTTP Listener is a connector that listens to HTTP requests. 
You can configure an HTTP Listener to start a flow when it receives specified requests.
```

> * Scheduler is a core component that starts a flow when a time-based condition is met. You can configure a Scheduler to start a flow at regular intervals, or you can specify a more flexible cron expression to start the flow.

Reviewer 2: ```suggestion
* Scheduler is a core component that starts a flow when a time-based condition is met. 
You can configure a Scheduler to start a flow at regular intervals, or you can specify a more flexible cron expression to start the flow.
```

> For more information about the HTTP Listener and the Scheduler component, see the xref:http/http-connector.adoc[HTTP Connector User Guide] and xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint (Trigger)].

Reviewer 2: ```suggestion
For more information about HTTP Listener and the Scheduler component, see the xref:http/http-connector.adoc[HTTP Connector User Guide] and xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint (Trigger)].
```

[[operations]]
=== Operations

Most connectors have operations that execute API calls or other actions on their associated system, database, or protocol. For example, you can use Anypoint Connector for Workday (Workday Connector) to create a position request in Workday or add a fund to the financial management service. Likewise, you can use Anypoint Connector for VMs (VM Connector) to consume messages from and publish messages to an asynchronous queue.

To see a list of operations for a connector, see the Reference Guide for that connector.

== Access Connectors and Related Assets

> Anypoint Exchange provides access to all publicly-available connector assets including connectors, templates, and examples.

Reviewer 2: GLOBALLY remove hyphens after compound adj when the first term ends in "ly":
```suggestion
Anypoint Exchange provides access to all publicly available connector assets including connectors, templates, and examples.
```

Author: I never knew that rule.

=== Connectors in Exchange

You can use Exchange as a starting point for discovering all or a subset of MuleSoft-built connectors.

For information about locating connectors in Exchange, see <<introduction/discover-add-connectors.adoc#Discover and Add Connectors>>.

=== Templates in Exchange

> Templates provide starting points for creating Studio projects. Many templates include one or more connector configurations.  For example, the S/4HANA Cloud to Salesforce Partner Migration template helps you migrate partners from an SAP S/4HANA Cloud to a Salesforce instance.

Reviewer 2: ```suggestion
Templates provide starting points for creating Anypoint Studio projects. Many templates include one or more connector configurations.  For example, the S/4HANA Cloud to Salesforce Partner Migration template helps you migrate partners from an SAP S/4HANA Cloud to a Salesforce instance.
```

> Once you configure a template in Flow Designer or Studio, you can customize it to meet your needs.

Reviewer 2: Once > After
GLOBAL, for localization (unless you really do mean "one time").

For information about locating and using templates, see <<introduction/discover-add-templates.adoc#Discover and Add Templates to Flow Designer or Studio>>.

=== Examples in Exchange

> Examples are apps that demonstrate how to handle a use case or solution. Many examples use one or more connectors to accomplish required tasks.

Reviewer 2: Can you think of a more specific term than "handle"? Again, it's a localization issue of sorts.
manage? process? use? ....

> For example, the Authenticating Salesforce using OAuth 2.0 example shows you how to connect to Salesforce using OAuth as the security protocol. It uses the Salesforce Connector to enable OAuth authentication before performing the integration process.

Reviewer 2: ```suggestion
For example, the Authenticating Salesforce Using OAuth 2.0 example shows you how to connect to Salesforce using OAuth as the security protocol. It uses Salesforce Connector to enable OAuth authentication before performing the integration process.
```

For information about locating and using examples, see <<introduction/discover-add-examples.adoc#Discover and Add Examples to Studio>>.

== Connector Support Categories

Anypoint Connectors have the following categories of support:

> * Select connectors are available to anyone. However, you must have an Anypoint Platform subscription to access MuleSoft support for these connectors.

Reviewer 2: ```suggestion
* Select connectors are available to anyone. 
However, you must have an Anypoint Platform subscription to access MuleSoft support for these connectors.
```

Author: Keeping this one together since it is so tightly linked with the first sentence.

* Premium connectors are available to licensed users only.

> * Mulesoft Certified connectors are developed by MuleSoft's partners and developer community and subsequently reviewed and certified by MuleSoft. Access and support for these connectors is through MuleSoft partners.

Reviewer 2: ```suggestion
* Mulesoft Certified connectors are developed by MuleSoft's partners and developer community and subsequently reviewed and certified by MuleSoft. 
Access to and support for these connectors is through MuleSoft partners.
```

Reviewer 2: The first two list items specify "available to." Should the third use the same term?

For more information about connector support categories, see the https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Product Versioning and Back Support Policy].

== Where to Get Additional Information

Access the Release Notes, User Guide, and Reference Guide for a connector to view connector-specific information:

> * Connector Release Notes provide compatibility information for a connector, information on new features, and a list of issues fixed in a release.

Reviewer 2: ```suggestion
* Connector Release Notes provide compatibility information for a connector, information about new features, and a list of issues fixed in a release.
```

* Connector User Guides provide instructions for configuring the associated connector and configuration examples.

> * Connector Reference Guides provide a list of supported input sources (if any) and operations for the associated connector. They also provide field descriptions for connector configurations and connection types.

Reviewer 2: ```suggestion
* Connector Reference Guides provide a list of supported input sources (if any) and operations for the associated connector. 
They also provide field descriptions for connector configurations and connection types.
```
